pyro/ops/stats.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+
+
+def _compute_chain_variance_stats(input):
+    # compute within-chain variance and variance estimator
+    N = input.size(1)
+    chain_mean = input.mean(dim=1)
+    var_between = chain_mean.var(dim=0)
+
+    chain_var = input.var(dim=1)
+    var_within = chain_var.mean(dim=0)
+    var_estimator = (N - 1) / N * var_within + var_between
+    return var_within, var_estimator
+
+
+def gelman_rubin(input):
+    """
+    Computes R-hat over chains of samples. The first two dimensions of `input`
+    is `C x N` where `C` is the number of chains and `N` is the number of samples.
+    It is required that `C >= 2` and `N >= 2`.
+
+    :param torch.Tensor input: the input tensor.
+    :returns torch.Tensor: R-hat of `input`.
+    """
+    assert input.dim() >= 2
+    assert input.size(0) >= 2
+    assert input.size(1) >= 2
+    var_within, var_estimator = _compute_chain_variance_stats(input)
+    return (var_estimator / var_within).sqrt()
+
+
+def split_gelman_rubin(input):
+    """
+    Computes R-hat over chains of samples. The first two dimensions of `input`
+    is `C x N` where `C` is the number of chains and `N` is the number of samples.
+    It is required that `C >= 2` and `N >= 4`.
+
+    :param torch.Tensor input: the input tensor.
+    :returns torch.Tensor: split R-hat of `input`.
+    """
+    assert input.dim() >= 2
+    assert input.size(1) >= 4
+    C, N_half = input.size(0), input.size(1) // 2
+    new_input = torch.stack([input[:, :N_half], input[:, -N_half:]], dim=1).reshape(
+        (2 * C, N_half) + input.shape[2:])
+    return gelman_rubin(new_input)
+
+
+def _fft_next_good_size(N):
+    # find the smallest number >= N such that the only divisors are 2, 3, 5
+    if N <= 2:
+        return 2
+    while True:
+        m = N
+        while m % 2 == 0:
+            m //= 2
+        while m % 3 == 0:
+            m //= 3
+        while m % 5 == 0:
+            m //= 5
+        if m == 1:
+            return N
+        N += 1
+
+
+def autocorrelation(input):
+    """
+    Computes the autocorrelation of samples. The last dimension is the number of
```

Comment:
I think it will be more natural to have the first dimension represent the different samples. This is how the `Empirical` distribution would represent it at least, and it will be nice to use it directly with these diagnostic functions.

Reply:
I'll add a parameter `dim` instead. By default, it will be `-1`.

Reply:
Sounds good! If its too much effort generalizing though, I wouldn't bother with it since there aren't too many parts of the code that would be calling these diagnostic functions (at least for the time being); so feel free to just change the default if that's easier.

Reply:
@neerajprad to make things consistence, I will follow your suggestion by setting `dim=0` by default. This is easier than the case with both chain and sample size. I just simply transpose dim with -1 for Fourier transform, then transpose the result back.